go-controller/pkg/ovn/default_network_controller.go

```diff
@@ -912,7 +887,9 @@ func (h *defaultNetworkControllerEventHandler) UpdateResource(oldObj, newObj int
 	case factory.EgressNodeType:
 		oldNode := oldObj.(*kapi.Node)
 		newNode := newObj.(*kapi.Node)
-
+		// Update node in zone cache; value will be true if node is local
+		// to this zone and false if its not
+		h.oc.eIPC.nodeZoneState.Store(newNode.Name, h.oc.isLocalZoneNode(newNode))
```

Comment:
So as we spoke offline on huddle, I have changed this up.

1. Now the new cache is done using our custom syncmap package.
2. All major paths are using the cache with a lock on the key: example: `addPodEgressIPAssignment`, `deletePodEgressIPAssignment` and `deleteEgressIPStatusSetup`.
3. I couldn't lock some parts like because it was tricky to do so :/ (open to ideas): 

A) In `deleteEgressIPStatusSetup` the part where we delete address-sets using `podIPs` is unsafe
```
// TODO(tssurya): Would be nice to lock the address-set deletion as well but its not that important
if len(podStatus.egressStatuses) == 0 && len(podStatus.standbyEgressIPNames) == 0 {
	// pod could be managed by more than one egressIP
	// so remove the podKey from cache only if we are sure
	// there are no more egressStatuses managing this pod
	klog.V(5).Infof("Deleting pod key %s from assignment cache", podKey)
	// delete the podIP from the global egressIP address set since its no longer managed by egressIPs
	if err := oc.deletePodIPsFromAddressSet(podIPs); err != nil {
		return fmt.Errorf("cannot delete egressPodIPs for the pod %s from the address set: err: %v", podKey, err)
	}
	delete(oc.eIPC.podAssignment, podKey)
} else if len(podStatus.egressStatuses) == 0 && len(podStatus.standbyEgressIPNames) > 0 {
	klog.V(2).Infof("Pod %s has standby egress IP %+v", podKey, podStatus.standbyEgressIPNames.UnsortedList())
	podStatus.egressIPName = "" // we have deleted the current egressIP that was managing the pod
	if err := oc.addStandByEgressIPAssignment(podKey, podStatus); err != nil {
		klog.Errorf("Adding standby egressIPs for pod %s with status %v failed: %v", podKey, podStatus, err)
		// We are not returning the error on purpose, this will be best effort without any retries because
		// retrying deleteEgressIPAssignments for original EIP because addStandByEgressIPAssignment failed is useless.
		// Since we delete statusToRemove from podstatus.egressStatuses the first time we call this function,
		// later when the operation is retried we will never continue down the loop
		// since we add SNAT to node only if this pod was managed by statusToRemove
		return nil
	}
}
```
B) In beginning of `addPodEgressIPAssignment` we use the cache to determine if we need to proceed with the ops:
```
// We need to proceed with add only under two conditions
// 1) egressNode present in at least one status is local to this zone
// (NOTE: The relation between egressIPName and nodeName is 1:1 i.e in the same object the given node will be present only in one status)
// 2) the pod being added is local to this zone
// TODO(tssurya): Check if this loop also needs to be locked for nodeZoneState key?
proceed := false
for _, status := range statusAssignments {
	isLocalZoneNode, loadedEgressNode := oc.eIPC.nodeZoneState.Load(status.Node)
	if loadedEgressNode && isLocalZoneNode {
		proceed = true
		break
	}
}
if !proceed && !oc.isPodScheduledinLocalZone(pod) {
	return nil // nothing to do if none of the status nodes are local to this master and pod is also remote
}
```

Reply:
oh err and maybe I need to always lockKey before loading and unloading as well? not sure...I feel like since we hold lock on entire cache we should be safe?

Reply:
resolving this comment, fixed this.